### ISSUE TITLE
Panel switch to wrong language problem

### ISSUE THREAD

**Issue by jenstornell** [getkirby/panel#943]
https://forum.getkirby.com/t/panel-language-session-problem/4959

If a field contains javascript it's a risk that they can make the panel save in the wrong language in a multi language setup. I think it's very serious because of the frustration that comes with lost/overwritten content.

This issue is not new, I just don't think it has been addressed before.

**Fields that I know suffer from it:**

https://github.com/jenstornell/kirby-seo
https://github.com/jenstornell/kirby-boiler-plugin
https://github.com/jenstornell/kirby-boiler-field

### 1. Install Kirby multi languages

**In the Kirby Cli:**

```
kirby install bigbug --kit langkit
```

### 2. Install boiler field or below

You can install the boiler field to try it, link above. If not here is an even simpler version of the boiler field:

**It's a field, not a plugin. It's named `boiler.php`**

```php
<?php
class BoilerField extends BaseField {
	static public $fieldname = 'boiler';
	static public $assets = array(
		'js' => array(
			'script.js',
		),
	);

	public function input() {
		$html = '<input type="text" class="input" id="' . $this->id() . '" name="' . $this->name() . '" value="' . $this->value() . '">';
		return $html;
	}

	public function element() {
		$element = parent::element();
		$element->data('field', self::$fieldname);
		return $element;
	}

	public function routes() {
		return array(
			array(
				'pattern' => 'ajax/(:any)/(:any)',
				'method'  => 'get',
				'action' => function($var1, $var2) {
					return response::json( array( $var1, $var2 ) );
				}
			)
		);
	}
}
```

**script.js**

```js
(function($) {
	$.fn.boiler = function() {
		return this.each(function() {
			var field = $(this);
			var fieldname = 'boiler';

			if(field.data( fieldname )) {
				return true;
			} else {
				field.data( fieldname, true );
			}
			$.fn.ajax(fieldname);
		});
	};

	$.fn.ajax = function(fieldname) {
		var blueprint_key = $('[data-field="' + fieldname + '"]').find('[name]').attr('name');
		var base_url = window.location.href.replace('/edit', '/field') + '/' + blueprint_key + '/' + fieldname + '/ajax/';
		$.ajax({
			url: base_url + '1/value',
			type: 'GET',
			success: function(result) {
				console.log('Hello from Ajax!');
				console.log(result);
			}
		});
	};
})(jQuery);
```

### 3. Add to your blueprint

**default.yml**

```yml
title: Bug
fields:
  title:
    label: Title
    type:  text
  boiler:
    label: Bug
    type:  boiler
```

### 4. Run it

In the test above, when I save in DE it switches back to EN.

![bigbug](https://cloud.githubusercontent.com/assets/927011/19849380/37ab6e6a-9f53-11e6-80c9-fb6eccf16cb8.gif)

This issue in this particular case has to do with the ajax route, that it does not go to a language specific version of the panel and that makes it switch back to EN.

With my Kirby SEO it's still unknown why the same thing happends there.

## In conclusion

For me this proves that the panel is quite fragile. A field can this easily break the save behavior. I want it to be solid as a rock.

**Suggestions:**

- Keep `en` or whatever language prefix in the url to never accidently switch away from it.
- Same as above with a get variable.
- Keep it in a hidden field, not in memory.

That would mean that I can't switch language by accedent from an ajax route or anything else. I will no longer be worried that my content is lost when using multiple languages.

**Comment by martinthiemann** [comment]
Yes please. I hope the potential solution will solve our panel problems, the symptom of overwriting files is the same.

**Comment by jonathan-reisdorf** [comment]
Just a very unconfirmed, random idea: Maybe the bug is somehow related to the POST calls that are being made to `/keep`? These calls are a feature to ensure that non-saved data is not lost when the user leaves the page and comes back later.

**Comment by lukasbestle** [comment]
https://forum.getkirby.com/t/really-strange-behavior-that-makes-kirby-save-to-wrong-content-file/5786?u=lukasbestle

**Comment by lukasbestle** [comment]
Fixed by #1125.
